### ISSUE TITLE
Dynamic rules based on request payload

### ISSUE THREAD

**Issue by rf-0** [scale-tone/ThrottlingTroll#32]
Hi,

First off, thanks for this amazing lib. Much prefer it to the AspNetCoreRateLimit and Microsoft.AspNetCore.RateLimiting.

My requirement:
As the title says, I'm trying to apply rules, per request, based on some payload. For example, lets say my project is a single public api endpoint. To call this endpoint you need an api key. This api key is mapped to a plan (ie, basic, pro or enterprise). When a request comes in, i want to do a lookup (redis) of the plan for this api key and based on that plan, i want to apply a different limit method, so for example, if basic 100 requests per minute, if pro 1_000, if enterprise 10_000 etc

I looked around and I did not find a way to do this using this library. Is it possible?

TLDR: Cant dynamically set rules/limits during each request based on some logic.

Thanks in advance!

**Comment by scale-tone** [comment]
Thanks @rf-0 !

> I looked around and I did not find a way to do this using this library. Is it possible?

Absolutely. We have [CostExtractors](https://github.com/ThrottlingTroll/ThrottlingTroll/wiki/150.-%5BIngress%5D-Personalized-rate-limiting#using-costextractors) precisely for that.

You set the total limit to 10000 coins, and then a basic chap should be charged 100 coins per request, a pro guy should pay 10 coins and an enterprise fellow pays just 1.

[Here's a code example](https://github.com/ThrottlingTroll/ThrottlingTroll-AspDotNetCore-Samples/blob/6374875417a8336eac0ba0f900557bab76bc6b51/ThrottlingTrollSampleWeb/Program.cs#L241).

**Comment by scale-tone** [comment]
No, no, IdentityIdExtractors and CostExtractors are synchronous on purpose, which purpose is not to encourage folks making network roundtrips as part of them. Remember, they're called as part of every request processing.

Yes, the best would be to cache your api-key map in HttpContext (periodically reload, if needed) and use it from there.

**Comment by scale-tone** [comment]
Forgot to mention that you would, of course, also need to provide an [IdentityIdExtractor](https://github.com/ThrottlingTroll/ThrottlingTroll/wiki/150.-%5BIngress%5D-Personalized-rate-limiting#extracting-and-using-identityids), so that this big limit is applied on a per api-key basis (and not the same limit for all active clients).
